zebra-network/src/meta_addr.rs

```diff
@@ -422,19 +398,90 @@ impl MetaAddr {
         self.last_failure
     }
 
-    /// Is this address a directly connected client?
-    pub fn is_direct_client(&self) -> bool {
-        match self.last_connection_state {
-            Responded => !self.services.contains(PeerServices::NODE_NETWORK),
-            NeverAttemptedGossiped | NeverAttemptedAlternate | Failed | AttemptPending => false,
+    /// Have we had any recently messages from this peer?
+    ///
+    /// Returns `true` if the peer is likely connected and responsive in the peer
+    /// set.
+    ///
+    /// [`constants::LIVE_PEER_DURATION`] represents the time interval in which
+    /// we should receive at least one message from a peer, or close the
+    /// connection. Therefore, if the last-seen timestamp is older than
+    /// [`constants::LIVE_PEER_DURATION`] ago, we know we should have
+    /// disconnected from it. Otherwise, we could potentially be connected to it.
+    pub fn was_recently_live(&self) -> bool {
+        if let Some(last_response) = self.last_response {
+            if let Some(elapsed) = last_response.elapsed() {
+                elapsed
+                    <= constants::LIVE_PEER_DURATION
+                        .as_secs()
+                        .try_into()
+                        .expect("unexpectedly large constant")
+            } else {
+                info!(last_response = ?self.last_response,
+                      now = ?DateTime32::now(),
+                      "unexpected future response time: assuming peer is live. Hint: has the system clock changed?"
+                );
+                // future times should be considered live
+                true
+            }
+        } else {
+            // if there is no response, it can't possibly be live
+            false
         }
     }
 
-    /// Is this address valid for outbound connections?
-    pub fn is_valid_for_outbound(&self) -> bool {
-        self.services.contains(PeerServices::NODE_NETWORK)
-            && !self.addr.ip().is_unspecified()
-            && self.addr.port() != 0
+    /// Have we recently attempted an outbound connection to this peer?
+    ///
+    /// Returns `true` if this peer was recently attempted, or has a connection
+    /// attempt in progress.
+    pub fn was_recently_attempted(&self) -> bool {
+        self.last_attempt
+            // `now` should always be later than `last_attempt`, except for tests
+            // so we can call future attempts "not recent" to simplify the code
+            .map(|last_attempt| Instant::now().checked_duration_since(last_attempt))
+            .flatten()
```

Comment:
Not really sure if this is a good idea, but the comment and `flatten` could be removed by having:

```suggestion
            .map(|last_attempt| Instant::now().checked_duration_since(last_attempt).unwrap_or(0))
```

Reply:
This is a good idea, but there's an even simpler refactor:
`Instant::now().checked_duration_since(last_attempt).unwrap_or(0)` is the same as 
`Instant::now().saturating_duration_since(last_attempt)`.

See commit 239fb70e.

The first line of the comment needs to stay, because it explains why we don't just use `elapsed` and panic here.

FYI if you're looking up the docs, there's a typo in `saturating_duration_since`: https://github.com/tokio-rs/tokio/pull/3864

Reply:
Nice! And good catch with the typo :tada: